packages/shared/src/time_control/time_control_systems.ts

```diff
@@ -0,0 +1,258 @@
+      state.periodsRemaining
+    })`;
+    if (state.mainTimeRemainingMS > 0) {
+      return `${msToTime(state.mainTimeRemainingMS)} + ${periodTimeString}`;
+    }
+    return periodTimeString;
+  }
+}
+
+class SimpleTimeControl
+  implements TimeControl<ITimeControlConfig, IBasicTimeState>
+{
+  initialState(config: ITimeControlConfig) {
+    return {
+      remainingTimeMS: config.mainTimeMS,
+    };
+  }
+
+  elapse(elapsedMS: number, state: IBasicTimeState) {
+    return { remainingTimeMS: state.remainingTimeMS - elapsedMS };
+  }
+
+  renew(_state: IBasicTimeState, config: ITimeControlConfig) {
+    return { remainingTimeMS: config.mainTimeMS };
+  }
+
+  msUntilTimeout(state: IBasicTimeState): number {
+    return state.remainingTimeMS;
+  }
+
+  timeString(state: IBasicTimeState): string {
+    return msToTime(state.remainingTimeMS);
+  }
+}
+
+export interface ICanadianTimeState {
+  mainTimeRemainingMS: number;
+  periodTimeRemainingMS: number;
+  renewalCountdown: number;
+}
+export interface ICanadianTimeConfig extends ITimeControlConfig {
+  type: TimeControlType.Canadian;
+  numPeriods: number;
+  periodTimeMS: number;
+}
+
+class CanadianTimeControl
+  implements TimeControl<ICanadianTimeConfig, ICanadianTimeState>
+{
+  initialState(config: ICanadianTimeConfig) {
+    return {
+      mainTimeRemainingMS: config.mainTimeMS,
+      periodTimeRemainingMS: config.periodTimeMS,
+      renewalCountdown: config.numPeriods,
+    };
+  }
+
+  elapse(elapsedMS: number, state: ICanadianTimeState) {
+    state = { ...state };
+    if (state.mainTimeRemainingMS) {
+      if (state.mainTimeRemainingMS >= elapsedMS) {
+        state.mainTimeRemainingMS -= elapsedMS;
+        return state;
+      }
+      elapsedMS -= state.mainTimeRemainingMS;
+      state.mainTimeRemainingMS = 0;
+    }
+    state.periodTimeRemainingMS -= elapsedMS;
+
+    return state;
+  }
+
+  renew(state: ICanadianTimeState, config: ICanadianTimeConfig) {
+    state = { ...state };
+    if (state.mainTimeRemainingMS > 0) {
+      // periods don't need to be renewed during main time, and we don't want
+      // to touch the countdown
+      return state;
+    }
+    state.renewalCountdown--;
+    if (state.renewalCountdown === 0) {
+      // reset
+      state.renewalCountdown = config.numPeriods;
+      state.periodTimeRemainingMS = config.periodTimeMS;
+    }
+    return state;
+  }
+
+  msUntilTimeout(state: ICanadianTimeState): number {
+    return state.mainTimeRemainingMS + state.periodTimeRemainingMS;
+  }
+
+  timeString(state: ICanadianTimeState): string {
+    const periodTimeString = `${msToTime(state.periodTimeRemainingMS)}/${
+      state.renewalCountdown
+    }`;
+    if (state.mainTimeRemainingMS > 0) {
+      return `${msToTime(state.mainTimeRemainingMS)} + ${periodTimeString}`;
+    }
+    return periodTimeString;
+  }
+}
+
+// Convenience function for when the state doesn't change at the end of a round
+function nullRenewal<T>(state: T) {
```

Comment:
nit: I believe T needs to be an iterable for this to work (or?). We could e.g. say
`nullRenewal<T extends object>`